### PR TITLE
Add Broker reconciler

### PR DESCRIPTION
**Pull request by pierDipi**
## Proposed Changes

- Add Broker reconciler
- Add a base reconciler for both Broker and Trigger reconcilers.

### CODE REVIEW

File: control-plane/pkg/reconciler/broker/broker.go [hunk 1]
```diff
@@ -19,39 +19,305 @@ package broker
...
-	)
+func incrementVolumeGeneration(generation uint64) uint64 {
+	return (generation + 1) % (math.MaxUint64 - 1)
+}
+
+func (r *Reconciler) createTopic(broker *eventing.Broker) (string, error) {
+	topic := topic(broker)
+	topicDetail := r.topicDetailFromBrokerConfig(broker)
+
+	r.KafkaClusterAdminLock.Lock()
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/broker.go:209]: Why we need locking here? KafkaClusterAdmin is not thread safe?

File: control-plane/pkg/reconciler/broker/broker.go [hunk 2]
```diff
@@ -19,39 +19,305 @@ package broker
...
+	defer r.KafkaClusterAdminLock.Unlock()
+
+	return topic, r.KafkaClusterAdmin.CreateTopic(topic, topicDetail, true)
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/broker.go:212]: `CreateTopic` returns an error if the topic is already there: https://github.com/knative/eventing-contrib/blob/master/kafka/channel/pkg/reconciler/controller/kafkachannel.go#L444

File: control-plane/pkg/reconciler/broker/broker.go [hunk 3]
```diff
@@ -19,39 +19,305 @@ package broker
...
-	return reconciler.NewEvent(
-		corev1.EventTypeNormal,
-		brokerReconciled,
-		fmt.Sprintf(`%s reconciled: "%s/%s"`, broker, namespace, name),
-	)
+func incrementVolumeGeneration(generation uint64) uint64 {
+	return (generation + 1) % (math.MaxUint64 - 1)
+}
+
+func (r *Reconciler) createTopic(broker *eventing.Broker) (string, error) {
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/broker.go:205]: Can you move the functions to manage topics in another file? it may help a bit the readability

File: control-plane/pkg/reconciler/broker/broker.go [hunk 4]
```diff
@@ -19,39 +19,305 @@ package broker
...
+	r.KafkaDefaultTopicDetailsLock.Lock()
+	defer r.KafkaDefaultTopicDetailsLock.Unlock()
+
+	r.KafkaDefaultTopicDetails = topicDetail
+}
+
+func findBroker(brokersTriggers *coreconfig.Brokers, broker *eventing.Broker) int {
+	// Find broker in brokersTriggers.
+	brokerIndex := noBroker
+	for i, b := range brokersTriggers.Broker {
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/broker.go:294]: Can you index the brokers with the UID directly more than storing them as array? If that requires a change in the proto schema, then do it in another PR, opening an issue for that

**Comment by slinkydeveloper** [comment]
Ok I pushed reviewed while i was still reviewing it, another round coming

File: control-plane/pkg/reconciler/broker/broker.go (hunk 1 above)
Comment (slinkydeveloper) [reply control-plane/pkg/reconciler/broker/broker.go:209]: Don't you create it only one time at bootstrap time?

Comment (slinkydeveloper) [reply control-plane/pkg/reconciler/broker/broker.go:209]: No ok I see now.

File: control-plane/pkg/reconciler/broker/controller.go [hunk 5]
```diff
@@ -18,48 +18,97 @@ package broker
...
-		func(obj interface{}) {
-			if trigger, ok := obj.(*eventing.Trigger); ok {
-				impl.EnqueueKey(types.NamespacedName{
-					Namespace: trigger.Namespace,
-					Name:      trigger.Spec.Broker,
-				})
-			}
-		},
-	))
+	watcher.Watch(configs.GeneralConfigMapName, func(configMap *corev1.ConfigMap) {
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/controller.go:85]: Can you put this handler directly in the reconciler code?

File: control-plane/pkg/reconciler/broker/controller.go [hunk 6]
```diff
@@ -18,48 +18,97 @@ package broker
...
+
+		numPartitionsStr := configMap.Data[defaultTopicNumPartitionConfigMapKey]
+		replicationFactorStr := configMap.Data[defaultTopicReplicationFactorConfigMapKey]
+		bootstrapServers := configMap.Data[bootstrapServersConfigMapKey]
+
+		logger := logging.FromContext(ctx)
+
+		numPartitions, err := strconv.Atoi(numPartitionsStr)
```
Comment (slinkydeveloper) [inline control-plane/pkg/reconciler/broker/controller.go:93]: If you look at https://github.com/knative/eventing-contrib/blob/master/kafka/channel/pkg/utils/util.go#L65, there is a way to read from config map and parse it without invoking manually Atoi etc

File: control-plane/config/500-controller.yaml [hunk 7]
```diff
@@ -30,10 +30,22 @@ spec:
...
+              value: knative-eventing
+            - name: BROKERS_TRIGGERS_CONFIG_MAP_NAME
+              value: kafka-broker-brokers-triggers
             - name: CONFIG_LOGGING_NAME
               value: config-logging
+            - name: CONFIG_OBSERVABILITY_NAME
+              value: config-observability
             - name: METRICS_DOMAIN
               value: knative.dev/eventing
+            - name: FORMAT
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/config/500-controller.yaml:43]: Sure.

File: control-plane/pkg/reconciler/base/reconciler.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+		"unmarshalling configmap",
+		zap.String("format", r.Format),
+	)
+
+	// determine unmarshalling strategy
+	switch r.Format {
+	case PROTOBUF:
+		err = proto.Unmarshal(brokersTriggersRaw, brokersTriggers)
+	case JSON:
+		err = json.Unmarshal(brokersTriggersRaw, brokersTriggers)
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/base/reconciler.go:80]: Yes, it does.

File: control-plane/pkg/reconciler/base/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+		logger.Debug(
+			"update dispatcher pod annotation",
+			zap.String("pod", fmt.Sprintf("%s/%s", p.Namespace, p.Name)),
+			zap.Uint64("volumeGeneration", volumeGeneration),
+		)
+
+		// do not update cache copy
+		pod := p.DeepCopy()
+		p = nil
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/base/reconciler.go:138]: My intention was to "disable" that variable so that misuses won't be possible.
I think it's not that difficult to use `p` instead of `pod` and then that bug is particularly difficult to find. If you have a better idea, let me know.

File: control-plane/pkg/reconciler/broker/broker.go [hunk 10]
```diff
@@ -19,39 +19,305 @@ package broker
...
+	coreconfig "knative.dev/eventing-kafka-broker/control-plane/pkg/core/config"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/log"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/base"
 )
 
 const (
-	broker           = "Broker"
-	brokerReconciled = broker + "Reconciled"
+	// topic prefix - (topic name: knative-<broker-namespace>-<broker-name>)
+	TopicPrefix = "knative-"
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:41]: This requires changes to the data plane as well as to the documentation, can we do it in a follow-up?

File: control-plane/pkg/reconciler/broker/broker.go [hunk 11]
```diff
@@ -19,39 +19,305 @@ package broker
...
+	Recorder record.EventRecorder
 }
 
 func (r *Reconciler) ReconcileKind(ctx context.Context, broker *eventing.Broker) reconciler.Event {
 
-	r.logger.Debug("reconciling Broker", zap.Any("broker", broker))
+	logger := logger(ctx, broker)
 
-	return reconciledNormal(broker.Namespace, broker.Name)
+	logger.Debug("reconciling broker")
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:66]: I'm applying this suggestion to all logging, is that right? Or do you want to do this just for this one?

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply control-plane/pkg/reconciler/broker/broker.go:66]: To all logging

**Comment by pierDipi** [comment]
> Maybe I lost it in the process, but before triggering the update with UpdateDispatcherPodsAnnotation do you assert that the data plane config map actually changed?

I'm not keeping track of the previous state (should I?), anyway on the data plane we are idempotent, so nothing breaks.

**Comment by pierDipi** [comment]
Ok, I left unresolved configurations related comments that we can reconcile with `eventing-contrib` later.

File: control-plane/pkg/reconciler/base/reconciler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply control-plane/pkg/reconciler/base/reconciler.go:138]: why don't you just reuse p without creating a new variable?

File: control-plane/pkg/reconciler/broker/broker.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply control-plane/pkg/reconciler/broker/broker.go:43]: ok open an issue for that

**Comment by slinkydeveloper** [comment]
> I'm not keeping track of the previous state (should I?), anyway on the data plane we are idempotent, so nothing breaks.

Sounds good

File: control-plane/pkg/reconciler/broker/controller.go (hunk 5 above)
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/controller.go]: Done

File: control-plane/pkg/reconciler/base/reconciler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/base/reconciler.go:138]: Make sense.

File: control-plane/pkg/reconciler/broker/broker.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:43]: https://github.com/knative-sandbox/eventing-kafka-broker/issues/32

File: control-plane/config/500-controller.yaml [hunk 12]
```diff
@@ -30,10 +30,22 @@ spec:
...
+            - name: BROKERS_TRIGGERS_CONFIG_MAP_NAME
+              value: kafka-broker-brokers-triggers
             - name: CONFIG_LOGGING_NAME
               value: config-logging
+            - name: CONFIG_OBSERVABILITY_NAME
+              value: config-observability
             - name: METRICS_DOMAIN
               value: knative.dev/eventing
+            - name: DATA_PLANE_CONFIG_FORMAT
+              value: JSON
```
Comment (slinkydeveloper) [inline control-plane/config/500-controller.yaml:44]: you didn't made lowercase here

**Comment by pierDipi** [comment]
sonatype breaks every day :sob: 

https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka-broker/31/pull-knative-sandbox-eventing-kafka-broker-integration-tests/1281523450262851584#1:build-log.txt%3A1387

https://testgrid.knative.dev/eventing-kafka-broker#continuous

/retest

**Comment by slinkydeveloper** [comment]
@pierDipi I feel you ahahah

**Comment by pierDipi** [comment]
again? :sob: 

https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka-broker/31/pull-knative-sandbox-eventing-kafka-broker-integration-tests/1281527853501911040#1:build-log.txt%3A1387

/retest

**Comment by pierDipi** [comment]
Now, it seems to be happy.

**Comment by pierDipi** [comment]
rebased following: https://github.com/knative-sandbox/eventing-kafka-broker/pull/35
